File: 1539. Kth Missing Positive Number/FindKthPositive.py
from typing import List


class Solution:
    
    def findLessOrEqualToTarget(self, target, nums):
        
        lo = 0
        hi = len(nums)-1
        ans = len(nums)
        
        while lo <= hi:
            mid = (lo+hi)//2
            if nums[mid] > target:
                ans = mid
                hi = mid-1
            else:
                lo = mid+1
                
        return ans
    
    
    
    # O(logn squared) time,
    # O(1) space,
    # Approach: binary search,
    def findKthPositive(self, arr: List[int], k: int) -> int:
        lo, hi = 1, arr[-1]+k
        ans = 1
        
        while lo <= hi:
            mid = (lo+hi)//2
            less_or_equal = self.findLessOrEqualToTarget(mid, arr)
            if mid == k + less_or_equal:
                ans = mid
                hi = mid-1
            elif mid < k + less_or_equal:
                lo = mid+1
            else:
                hi = mid-1
                
        return ans
    

    # O(n) time,
    # O(1) space,
    # Approach: two pointers
    def findKthPositive2(self, arr: List[int], k: int) -> int:
        
        n = 1
        i = 0
        while k > 0 and i < len(arr):
            if n == arr[i]:
                i += 1
                n += 1
            else:
                n += 1
                k -= 1
                
        return n + k-1

File: 1539. Kth Missing Positive Number/find_kth_positive.rs
impl Solution {
    
    fn count_less_or_equal_target(target: i32, arr: &Vec<i32>) -> i32 {
        
        let mut lo = 0;
        let mut hi = arr.len()-1;
        let mut ans = arr.len();
        
        while hi <= arr.len() && lo <= hi {
            let mid = lo + (hi-lo)/2;
            if arr[mid] > target {
                ans = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        
        return ans as i32;
    }
    
    
    // O(logn squared) time,
    // O(1) space,
    // Approach: binary search, 
    pub fn find_kth_positive(arr: Vec<i32>, k: i32) -> i32 {
        
        let mut lo = 1;
        let mut hi = arr[arr.len()-1]+k;
        let mut ans = 1;
        
        while lo <= hi {
            let mid = lo + (hi-lo)/2;
            let less_count = Solution::count_less_or_equal_target(mid, &arr);
            if k + less_count == mid {
                ans = mid;
                hi = mid-1;
            } else if k + less_count > mid {
                lo = mid+1;
            } else {
                hi = mid-1;
            }
        }
        
        return ans;
    }
}